### PR TITLE
Keep backward compatibility for 'node.Spec.Unschedulable'.

### LINKED ISSUE #68899: Failing Test : persistent-volume-upgrade [sig-storage]
### Failing Job
---
[sig-release-master-upgrade#gce-new-master-upgrade-master](https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-master)

### Gubernator Logs
---
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1778

### Error
---
```
Expected error:
    <*errors.errorString | 0xc420a4b980>: {
        s: "failed to get logs from pvc-tester-j4c4l for write-pod: the server does not allow this method on the requested resource (get pods pvc-tester-j4c4l)",
    }
    failed to get logs from pvc-tester-j4c4l for write-pod: the server does not allow this method on the requested resource (get pods pvc-tester-j4c4l)
not to have occurred

k8s.io/kubernetes/test/e2e/framework.(*Framework).testContainerOutputMatcher(0xc42097b340, 0x443f662, 0xf, 0xc42026a000, 0x0, 0xc420090e98, 0x1, 0x1, 0x45d0920)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2350 +0x216
k8s.io/kubernetes/test/e2e/framework.(*Framework).TestContainerOutput(0xc42097b340, 0x443f662, 0xf, 0xc42026a000, 0x0, 0xc420090e98, 0x1, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:475 +0x8a
k8s.io/kubernetes/test/e2e/upgrades/storage.(*PersistentVolumeUpgradeTest).testPod(0x6c17120, 0xc42097b340, 0x449c1c2, 0x20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/upgrades/storage/persistent_volumes.go:100 +0x10e
k8s.io/kubernetes/test/e2e/upgrades/storage.(*PersistentVolumeUpgradeTest).Test(0x6c17120, 0xc42097b340, 0xc42151eae0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/upgrades/storage/persistent_volumes.go:82 +0x8f
k8s.io/kubernetes/test/e2e/lifecycle.(*chaosMonkeyAdapter).Test(0xc4217d7640, 0xc421382b80)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/lifecycle/cluster_upgrade.go:447 +0x1da
k8s.io/kubernetes/test/e2e/lifecycle.(*chaosMonkeyAdapter).Test-fm(0xc421382b80)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/lifecycle/cluster_upgrade.go:516 +0x34
k8s.io/kubernetes/test/e2e/chaosmonkey.(*chaosmonkey).Do.func1(0xc421382b80, 0xc42157c8a0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/chaosmonkey/chaosmonkey.go:89 +0x76
created by k8s.io/kubernetes/test/e2e/chaosmonkey.(*chaosmonkey).Do
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/chaosmonkey/chaosmonkey.go:86 +0xa8
```

### Triage
---
This test is a persistent fail in this one test bucket only, and aside from one flake it appears to consistently pass everywhere else in sig-release-master-upgrade and in sig-release-1.12-all.

/kind bug
/priority failing-test
/sig storage
@kubernetes/sig-storage-bugs 
/assign @msau42 
/cc @saad-ali @childsb 
/milestone v1.12

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #68899 

**Release note**:
```release-note
Fixes issue [#68899](https://github.com/kubernetes/kubernetes/issues/68899) where pods might schedule on an unschedulable node.
```

### CODE REVIEW

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -1470,7 +1470,15 @@ func CheckNodeUnschedulablePredicate(pod *v1.Pod, meta algorithm.PredicateMetada
 		return false, []algorithm.PredicateFailureReason{ErrNodeUnknownCondition}, nil
 	}
 
-	if nodeInfo.Node().Spec.Unschedulable {
+	// If pod tolerate unschedulable taint, it's also tolerate node.Spec.Unschedulable.
+	unschedulable := !v1helper.TolerationsTolerateTaint(pod.Spec.Tolerations, &v1.Taint{
```
Comment (liggitt) [inline pkg/scheduler/algorithm/predicates/predicates.go]: the var name makes this hard to follow, as does combining info about the node and info about the pod into a single variable below. would something like this be clearer?

```go
podToleratesUnschedulable := v1helper.TolerationsTolerateTaint(...)`
if nodeInfo.Node().Spec.Unschedulable && !podToleratesUnschedulable {
  return false, []algorithm.PredicateFailureReason{ErrNodeUnschedulable}, nil
}
```

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -1470,7 +1470,15 @@ func CheckNodeUnschedulablePredicate(pod *v1.Pod, meta algorithm.PredicateMetada
...
+		Key:    algorithm.TaintNodeUnschedulable,
+		Effect: v1.TaintEffectNoSchedule,
+	})
+	// TODO (k82cn): deprecates `node.Spec.Unschedulable` in 1.13.
```
Comment (liggitt) [inline pkg/scheduler/algorithm/predicates/predicates.go]: it's unclear what is being deprecated here, or what action should be taken in 1.13

Comment (k82cn) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Update related doc to depreciate it.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 1 above)
Comment (k82cn) [reply pkg/scheduler/algorithm/predicates/predicates.go]: done

**Review by tpepper — CHANGES_REQUESTED** [review]
Given all of the conversation in https://github.com/kubernetes/kubernetes/issues/68899 and the nature of the problem here, I think this needs a better commit message still and might even deserve a release note under "Known Issues" or SIGs Scheduling/Node to at least remind operators what's changed here and the taints and tolerations they need to be considering across updates in this version range.

**Comment by k82cn** [comment]
> I think this needs a better commit message still and might even deserve a release note

There's info in the description; no behaviour changed, so no release note here.

**Comment by k82cn** [comment]
> Plus we do have a bunch of "fixes issue where..." type release notes.

Such kind of release node is ok to me.

**Comment by krmayankk** [comment]
will `kubectl cordon` be implemented using unschedulable taint instead of node.Spec.Unschedulable ?

**Comment by tpepper** [comment]
>> Plus we do have a bunch of "fixes issue where..." type release notes.
>
> Such kind of release node is ok to me.

OK then we need confirmation this follows [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) which I'm not sure is actually true, the relnote wording specified below, a decision on whether to put this is the Node or Scheduling notes section, and the deprecation issue opened against 1.13.

Possible relnote:
```
Fixes issue [#68899](https://github.com/kubernetes/kubernetes/issues/68899) where pods might schedule on an incorrectly tainted node. Deprecation notice: The alpha feature "Critical Pods" has been replaced and `node.Spec.Unschedulable` will be removed in 1.13 as per issue #XXXXXX.
```

**Comment by liggitt** [comment]
The api field will definitely not be removed from the v1 api. It's not entirely clear to me what behavior can be deprecated in v1. If a client stops setting it and just tries to set the taint, the controller will remove the taint automatically, right?

If kubectl wanted to begin simultaneously setting the unscheduleable field and taint in cordon (and removing them both in uncordon), that could maybe be ok, but would need to think about that a little more.

**Comment by tpepper** [comment]
I really want to see this PR finished off today so we get final confirmation on long running tests over the coming days head of 1.12.  This is an important compatibility issue to resolve thoroughly.

**Comment by tpepper** [comment]
...and this is the only PR I'm holding the 1.12 release on at this point.

**Comment by neolit123** [comment]
looks like anyone can LGTM this at this point.
(it is approved, has labels and milestone)

are further changes required here?

**Comment by neolit123** [comment]
/hold
just in case.

**Comment by tpepper** [comment]
Chatting with @bsalamat it seems I've misread the intent here.  Clarifying based on that:

* release note suggestion: ```Fixes issue [#68899](https://github.com/kubernetes/kubernetes/issues/68899) where pods might schedule on an incorrectly tainted node.```
* Bobby will open an issue to track a future deprecation.
* Deprecation announcement would be targeted for 1.13, with actual removal then being at least a year out given this is v1 stable content.

**Review by bsalamat — APPROVED** [review]
/lgtm

**Review by tpepper — APPROVED** [review]

**Comment by k82cn** [comment]
> will kubectl cordon be implemented using unschedulable taint instead of node.Spec.Unschedulable ?

Eventually, Yes; but we need to follow deprecation policy to do that.

> If kubectl wanted to begin simultaneously setting the unscheduleable field and taint in cordon (and removing them both in uncordon), 
> that could maybe be ok, but would need to think about that a little more.

Agree, we need to think about it more; we can follow up this in #69010 .

File: pkg/scheduler/algorithm/predicates/predicates_test.go [hunk 3]
```diff
@@ -4982,3 +4982,65 @@ func TestGetMaxVols(t *testing.T) {
...
+				},
+			},
+			fit: true,
+		},
+	}
+
+	for _, test := range testCases {
+		nodeInfo := schedulercache.NewNodeInfo()
+		nodeInfo.SetNode(test.node)
+		fit, _, err := CheckNodeUnschedulablePredicate(test.pod, nil, nodeInfo)
```
[the remark this answers is not in the document]
Comment (k82cn) [reply pkg/scheduler/algorithm/predicates/predicates_test.go]: Added it by `RegisterMandatoryFitPredicate` in this PR.
